pkg/raft/log_unstable.go

```diff
@@ -227,23 +227,26 @@ func (u *unstable) stableSnapTo(i uint64) {
 // restore resets the state to the given snapshot. It effectively truncates the
 // log after s.lastEntryID(), so the caller must ensure this is safe.
 func (u *unstable) restore(s snapshot) bool {
-	// All logs >= s.term are consistent with the log covered by the snapshot. If
-	// our log is such, disallow restoring a snapshot at indices below our last
+	mark := s.commitMark()
+	// All logs >= mark.Term are consistent with the log covered by the snapshot.
+	// If our log is such, disallow restoring a snapshot at indices below our last
 	// index, to avoid truncating a meaningful suffix of the log and losing its
 	// durability which could have been used to commit entries in this suffix.
 	//
 	// In this case, the caller must have advanced the commit index instead.
 	// Alternatively, we could retain a suffix of the log instead of truncating
 	// it, but at the time of writing the whole stack (including storage) is
 	// written such that a snapshot always clears the log.
-	if s.term <= u.term && s.lastIndex() < u.lastIndex() {
+	if mark.Term <= u.term && mark.Index < u.lastIndex() {
 		return false
 	}
-	// If s.term <= u.term, our log is consistent with the snapshot. Set the new
-	// accepted term to the max of the two so that u.term does not regress. At the
-	// time of writing, s.term is always >= u.term, because s.term == raft.Term
-	// and u.term <= raft.Term. It could be relaxed in the future.
-	term := max(u.term, s.term)
+	// If mark.Term <= u.term, our log is consistent with the snapshot. Set the
+	// new accepted term to the max of the two so that u.term does not regress. At
+	// the time of writing, mark.Term is always >= u.term, because mark.Term ==
+	// raft.Term and u.term <= raft.Term. It could be relaxed in the future.
```

Comment:
Good example.

> the snapshot will have index=lastIndex(recipient), term 3, and the unstable on the recipient will have term 4, no?

Yes, the snapshot will be at `entryID` = `{index:lastIndex, term:3}`, but the commit mark in this terminology is `{term:5, index:lastIndex}`.

So, at the recipient, who is at `LogMark.Term == 4` and, say, `Term=4`, we will overwrite the log with this newer-term committed state, and update `Term` to 5 and `LogMark.Term` to 5 (the `max` in the next line). Our log mark moves forward, and we're safe to regress to the snapshot's index as long as its last `entryID` does not match our log (which is checked up the stack in the caller). If it matches then there is a chance we have a suffix that must be retained.

---

Another example (impossible situation):

```
log (terms only)

        3 3 3 3   <- sender (all committed at term 3)
       /
3 3 3
       \
        4 4 4 4 4 4 4    <- recipient
```

In this case, the commit `mark` is `{term:3, index}`. Since these were committed at term 3, all term 4 logs must contain these entries. So we can't have this situation.